### PR TITLE
Fix gpus not being considered when counting allocatables

### PR DESCRIPTION
**Pull request by ElijahQuinones**
**Is this a bug fix or adding new feature?**
bug fix to address part of gh [#2105](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/issues/2105) gpus not being considered when counting allocatables.

**What is this PR about? / Why do we need it?**
We need this PR in order to fix the above referenced bug

**What testing is done?** 
Compared manually to [list of gpu instances](https://aws.amazon.com/ec2/instance-types/)
and created unit tests for instances mentioned in ticket and for a gpu instance with 4 gpus.

### CODE REVIEW

File: pkg/driver/node_test.go
Comment (AndrewSirenko) [inline pkg/driver/node_test.go:1220]: Nit: Would it be helpful to append this comment to the test name itself `g4dn.12xlarge_volume_attach_limit (4 GPUs, 1 InstanceStoreVolume)`, and adding the number of Instance Store Volumes instead of commenting? 

I'm thinking that this would be helpful if these unit tests fail. Developers can easily see test name for the test's assumptions, instead of stumbling across this comment. Furthermore the inclusion of InstanceStoreVolumes makes clear that GPUs aren't the only thing that uses up slot.

File: hack/generate-gpu-count-table.sh
Comment (AndrewSirenko) [inline hack/generate-gpu-count-table.sh:28]: Even if we can't combine gpu and InstanceStore queries into one, can we parallelize this inner loop? 

**This script goes from 4.25 minutes to 39 seconds with this change.** 

```
"${BIN}/aws" account list-regions --max-results 50 | jq -r '.Regions | map(.RegionName) | .[]' | while read REGION; do
     sleep 1
     get_gpus_for_region $REGION &
    done
```


The `&` "terminates an asynchronous command" such that it will run in background and run next command (next loop iteration) immediately.

The `sleep 1` sleeps for one second, which prevents the rate-limit issues @ConnorJC3 mentioned during demo. (Or atleast I'm not running into any limits locally) 

Same thing for the instance store generation function


Speedup proof:

```
./hack/generate-gpu-count-table.sh  30.72s user 4.08s system 88% cpu 39.507 total
./hack/generate-gpu-count-table.sh  36.98s user 5.20s system 16% cpu 4:16.95 total
```

I also diffed the results and found no difference.

File: pkg/driver/node_test.go
Comment (ElijahQuinones) [reply pkg/driver/node_test.go:1220]: agree will add in next rev

**Comment by rdpsin** [comment]
Should we do some manual testing by setting up a cluster with GPU instances?

File: pkg/cloud/volume_limits.go
Comment (AndrewSirenko) [inline pkg/cloud/volume_limits.go:500]: In a future PR, it would be nice if we could auto-generate these to prevent operator error (perhaps make a new makefile `pre-release` target that also updates sidecars). 

Here is the magic CLI incantation I came up that worked to auto-gen these locally for me if you want to alley-oop it. But I don't mind making the followup PR during the v1.34 release if you're swamped this sprint. 

If you wrap the actual map with `// CODEGEN_GPU START` and ``CODEGEN_GPU STOP` e.g.

```
var gpuInstanceGpus = map[string]int{
// CODEGEN_GPU START
...
// CODEGEN_GPU STOP
}
```

Then the following incantation when run with gnu-sed will update the `volume_limits.go` file in-place

```
ROOT_DIR="$(dirname "$(realpath "${BASH_SOURCE[0]}")")/../../"
BIN="$ROOT_DIR/bin"
VOLUME_LIMITS_FILE="$ROOT_DIR/pkg/cloud/volume_limits.go"

...

export GPU_LIMITS
GPU_LIMITS=$(get_all_gpus | sort | uniq)
sed -n -e "1,/CODEGEN_GPU START/ p" -e "/CODEGEN_GPU STOP/,$ p" "$VOLUME_LIMITS_FILE" -i
sed '/CODEGEN_GPU\ START/a $GPU_LIMITS' "$VOLUME_LIMITS_FILE" -i
envsubst < "$VOLUME_LIMITS_FILE" | sponge "$VOLUME_LIMITS_FILE"
```

^^ does the following:

1. Store the de-duplicated results of `get_all_gpus` in an exported environment variable
2. Deletes all text between the CODEGEN_GPU START & STOP comments
3. Adds the `$GPU_LIMITS` environment between the START and STOP comments which will be...
4. Substituted with the actual list of GPU Limits and written to `volume_limits.go` 

Note that a few things need to be cleaned up before PR (e.g. making sure the limits get pasted with appropriate tabbing) but it's good enough for me to use next release.  

Hope this was helpful

File: pkg/cloud/volume_limits.go
Comment (AndrewSirenko) [inline pkg/cloud/volume_limits.go:141]: Would you add a [Go Doc Comment](https://tip.golang.org/doc/comment) above this function that describes what it's doing? 

Something like:
```
// GetReservedSlotsForInstanceType calculates how many attachment slots are already used up by other devices. These devices include Instance Store Volumes and GPUs
```

It's good practice to provide a go doc comment for any exported (capitalized) name so that those who import the package can easily see what it is used for.

For example, IDEs like VsCode and GoLand are able to parse out these go docs to provide on-hover descriptions of the exported function/const.

**Comment by ElijahQuinones** [comment]
> Should we do some manual testing by setting up a cluster with GPU instances?

I have manually tested the limit outside of Kubernettes entirely by just spinning up a g4ad.xlarge instance and mannualy attaching ebs volumes until I hit the limit where all volumes afterwards are stuck in the attaching state this test concluded that 24 volumes could be attached to the g4ad.xlarge which complies with what the documentation provides.

starts at 28
-1 eni
-1 root volume
-1 gpu 
-1 instance store 
we also ens up with 24 

Additionally the unit tests for a g4ad.xlarge return the expected value of 24 slots.

however if we feel it would cover more bases I can run the test with a g4ad.xlarge cluster just to make sure there is no gotchas.

File: pkg/cloud/volume_limits.go
Comment (ConnorJC3) [reply pkg/cloud/volume_limits.go:500]: Suggest we hold off on this for now as this table will likely be dropped in the near future

Comment (ElijahQuinones) [reply pkg/cloud/volume_limits.go:500]: I agree with Connor here that we may want to hold off for now.However I  will add a ticket on the backlog to do this though as it will be a good improvement

**Review by ConnorJC3 — APPROVED** [review]
I've personally done enough manual testing to be confident it works

Needs squash otherwise lgtm

**Review by torredil — APPROVED** [review]
Thanks for including additional unit tests.
/lgtm

**Pull request merged**
